// @zappar/zapworks-cli/out/modules/zpp/src/zpputils/zipfile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZIPFileWriter = exports.ZIPFileReader = void 0;
const jszip = require("jszip");
const fs = require("fs-extra");
class ZIPFileReader {
    constructor(filename) {
        this._filename = filename;
    }
    async _getZIP() {
        if (this._zip)
            return await this._zip;
        let f = await fs.readFile(this._filename);
        this._zip = jszip.loadAsync(f);
        return await this._zip;
    }
    async getFile(filename) {
        let f = await this._getZIP();
        let zip = f.file(filename);
        if (!zip)
            throw new Error("Unable to load: " + filename);
        return await zip.async("nodebuffer");
    }
}
exports.ZIPFileReader = ZIPFileReader;
class ZIPFileWriter {
    constructor(filename) {
        this._filename = filename;
        this._zip = jszip();
    }
    async putFile(filename, contents) {
        this._zip.file(filename, contents, {
            compression: "DEFLATE"
        });
    }
    async save() {
        let buffer = await this._zip.generateAsync({ type: "nodebuffer" });
        await fs.writeFile(this._filename, buffer);
    }
}
exports.ZIPFileWriter = ZIPFileWriter;
